crates/ty_ide/src/docstring.rs

```diff
@@ -153,6 +149,128 @@ fn documentation_trim(docs: &str) -> String {
+    let mut block_indent = 0;
+    let mut in_doctest = false;
+    let mut starting_literal = false;
+    let mut in_literal = false;
+    let mut in_any_code = false;
+    for untrimmed_line in docstring.lines() {
+        // We can assume leading whitespace has been normalized
+        let mut line = untrimmed_line.trim_start_matches(' ');
+        let line_indent = untrimmed_line.len() - line.len();
+
+        // First thing's first, add a newline to start the new line
+        if !first_line {
+            // If we're not in a codeblock, add trailing space to the line to authentically wrap it
+            if !in_any_code {
+                output.push_str("  ");
+            }
+            // Only push newlines if we're not scanning for a real line
+            if !starting_literal {
+                output.push('\n');
+            }
+        }
+
+        // If we're in a literal block and we find a non-empty dedented line, end the block
+        if in_literal && line_indent < block_indent && !line.is_empty() {
+            in_literal = false;
+            in_any_code = false;
+            block_indent = 0;
+            output.push_str("```\n");
+        }
+
+        // We previously entered a literal block and we just found our first non-blank line
+        // So now we're actually in the litreal
+        if starting_literal && !line.is_empty() {
+            starting_literal = false;
+            in_literal = true;
+            in_any_code = true;
+            block_indent = line_indent;
+            // TODO: should we not be this aggressive? Let it autodetect?
+            output.push_str("\n```python\n");
+        }
+
+        // If we're not in a codeblock and we see something that signals a doctest, start one
+        if !in_any_code && line.starts_with(">>>") {
+            block_indent = line_indent;
+            in_doctest = true;
+            in_any_code = true;
+            // TODO: is there something more specific? `pycon`?
+            output.push_str("```python\n");
+        }
+
+        // If we're not in a codeblock and we see something that signals a literal block, start one
+        if !in_any_code && let Some(without_lit) = line.strip_suffix("::") {
+            let trimmed_without_lit = without_lit.trim();
+            if let Some(character) = trimmed_without_lit.chars().next_back() {
+                if character.is_whitespace() {
+                    // Remove the marker completely
+                    line = trimmed_without_lit;
+                } else {
+                    // Only remove the first `:`
+                    line = line.strip_suffix(":").unwrap();
+                }
+            } else {
+                // Delete whole line
+                line = trimmed_without_lit;
+            }
+            starting_literal = true;
+        }
+
+        // Add this line's indentation.
+        // We could subtract the block_indent here but in practice it's uglier
+        for _ in 0..line_indent {
```

Comment:
This won't work for code using tab indentation (which our formatter supports).

Reply:
(another thing handled by `documentation_trim`)